List feature

aggiewb opened the pull request.

aggiewb committed:
Add html, css, and js empty files

aggiewb committed:
Add css for body, h1, h2, and inputs

aggiewb committed:
Add basic structure w/ form to add items

aggiewb committed:
Add img for removing item off of list

aggiewb committed:
Add functions addListItems() and createImgElement() used

within addListItems()

aggiewb committed:
Add script tag for js, and removing h2 element

aggiewb committed:
Update li list-style-type, and set img margin-left

aggiewb committed:
Move imgElement variable with createImgElement() call inside

the for loop so that it creates a new image element on each array item
Add function removeListItem() using the X image, and add a click event listener on the image

aggiewb committed:
Add if conditional to return from addListItems() if the user

submits without entering an item
Add form reset after an li element has been added to ul
Move event.preventDefault() to end of addListItems() function

aggiewb committed:
Rename grocery list to shopping list

aggiewb committed:
Update colors

aggiewb committed:
Add p element for user warning

aggiewb committed:
Add if/else block to display user warning in p tag if user does

not enter an item

aggiewb committed:
Add button element which will clear all items out of a list

aggiewb committed:
Add button styling

aggiewb committed:
Change variable ulElement to global as multiple functions use

it
Add additional conditionals to for loop in addListItems()

aggiewb committed:
Add function createLiElementAppendTiUl() and add to

addListItems() passing in createXImgElement() and an item in the items array

aggiewb committed:
Add html footer element with js that set the innerHTML

aggiewb committed:
Add styling for footer and its child a tag element

aggiewb committed:
Update function name to appendItemToList() for easier

readability

BatemanVO commented on css/style.css:
```diff
@@ -0,0 +1,66 @@
+}
+
+input[type=submit]:hover, button:hover{
+    background-color: #FFFFFF;
```
If you're interested, a shorthand for hex values in CSS where all 6 characters is the same is to just use three characters - `#FFF` is shorthand for `#FFFFFF`.

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+var ulElement = document.querySelector('ul');
```
Since there is only one `<ul>` on this page, it makes sense to cache it in a variable called `ulElement` - but a more informative name might be something like `shoppingList` or `itemList` or something that better describes what the `ul` ultimately represents.

Additionally, we should wrap this code in an IIFE for better encapsulation and not setting unnecessary global variables.

One thing you should start using from here on is [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) - it's a simple `'use strict';` at the top of your file, or as the first line of your IIFE, that will head off certain issues (discussed in the linked MDN article).

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+
+function addListItems(event){
+    var form = event.target;
+    var items = form.querySelector('#items').value.split(',');
```
We could refactor this code to the following:
![Capture](https://user-images.githubusercontent.com/20998142/84095775-0538ba80-a9b5-11ea-8e34-740ad00969d0.PNG)

[The Array.prototype.map method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) declares a new Array, iterates over each element of the Array you called `.map` on, then pushes whatever your function returns for that given item to the new Array, then returns the new Array. In short, it passes each element of the original Array into your provided function, then transforms it to whatever your function returns.

[String.prototype.trim](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim) removes whitespace from the beginning and end of a String. So what we're ultimately saying is, "Take this String, split it on commas, then for each item, trim all the whitespace from the start and end."

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+
+    if(items[0].length > 0){
+        document.querySelector('#user-warning').textContent = "";
+        for(var i = 0; (i < items.length) && (items[i].length > 0) && (items[i] !== ' '); i++){
```
We could use another function similar to `.map` called [Array.prototype.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) that filters your Array for any items for which your provided function does not return true or a truthy value. So, if we were to add to the other code I suggested:

![Capture](https://user-images.githubusercontent.com/20998142/84096101-db33c800-a9b5-11ea-80b0-c308f4d079be.PNG)

So we'll remove all the excess whitespace, then additionally remove any elements that, with whitespace removed, have a length of 0. ~~Then we can remove the `if` statement altogether because if `nonEmptyItems.length` is 0, the `for` loop won't execute~~The conditional for the `for` loop is also much simpler to read and understand.

EDIT: Nevermind, you'll want to leave the `if` statement in - I didn't see the `else` that sets a user error message if they didn't enter any items. However, you can change it to `if (nonEmptyItems.length > 0)`.

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+function createXImgElement(){
+    var imgElement = document.createElement('img');
+    imgElement.setAttribute('src', 'media/x-icon.png');
+    imgElement.setAttribute('alt', 'A red x to indicate removal.');
```
We could expand this message a bit to say, "An X that can be clicked to remove this item from the list", to help indicate a bit more clearly to visually disabled people what the purpose of the X is.

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+        document.querySelector('#user-warning').textContent = "Please enter an item.";
+    }
+    event.preventDefault();
+    form.reset();
```
I didn't know about this method - nice find!

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+    return imgElement;
+}
+
+function appendItemToList(imgElement, item){
```
We could simplify this method a little by removing the `imgElement` parameter and instead changing line 30 to `liElement.appendChild(createXImgElement());`. Since we're always appending the same return value, there's no need for it to be a parameter - and since we only reference it on that one line, there's no need to cache it.

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+}
+
+function clearList(){
+    while(ulElement.hasChildNodes()){
```
Interesting find - I've been using `while (element.firstChild)` and relying on truthy values, but something that explicitly returns a Boolean is nice. I may look at swapping out logic at work with this method instead.

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+document.querySelector('button').addEventListener('click', clearList);
+
+var date = new Date();
+document.querySelector('footer').innerHTML = "Aggie Wheeler Bateman &copy; " +  date.getFullYear() + " <a href=\"https://www.aggiewheelerbateman.com\" target=\"_blank\">Personal Portfolio</a>";
```
One possible refactor we could do for this line to make it a little easier for the next developer to understand why the `<footer>` is empty in the markup, even though it's being styled, is change the HTML and JS to the following:
![Capture](https://user-images.githubusercontent.com/20998142/84096756-8ee98780-a9b7-11ea-9bea-9804b9fef844.PNG)

[The MDN article on innerHTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML) mentions that use of `innerHTML` often gets rejected by automated security review software, so removing its use is generally considered good practice. Additionally, it should be more efficient to have "static" markup in the HTML instead of dynamically adding it in JS - if we use this refactor, then we're only modifying the `textContent` of a single span with the only actually dynamic value.

BatemanVO commented on css/style.css:
```diff
@@ -0,0 +1,66 @@
+body {
+    background-color: #39393A;
```
You might be interested in investigating [CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties), also known as CSS variables. They work just like JS variables, allowing you to cache the same style in one place in case you want to make quick changes in the future without having to find every instance of a particular value. I notice this color is used in a couple places.

You don't have enough repeated values to really demand the use of custom properties here, but it might be something you're interested in in the future.

BatemanVO commented on css/style.css:
```diff
@@ -0,0 +1,66 @@
+
+a:hover {
+    color: #FFFFFF;
+}
```
I like that this CSS file is organized with the selectors generally in the order they appear on the page - it makes it easier to find where the rule for a given element is within the file.

aggiewb replied:
Thanks! I'm trying to find better ways to organize my css selectors.

BatemanVO commented on index.html:
```diff
@@ -0,0 +1,23 @@
+        <link href="css/style.css" rel="stylesheet">
+        <link href="https://fonts.googleapis.com/css2?family=Jost&family=Open+Sans&display=swap" rel="stylesheet"> 
+        <meta charset="UTF-8">
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
```
Good structuring of the `<head>` element - many of these metadata elements are often missed by developers.

aggiewb replied:
Thanks! I want to look into making a template for standard HTML structure.

BatemanVO commented on js/shopping-list.js:
```diff
@@ -0,0 +1,50 @@
+            appendItemToList(createXImgElement(), items[i]);
+        }
+    } else {
+        document.querySelector('#user-warning').textContent = "Please enter an item.";
```
We have mixed use of single and double-quotes for Strings in this file - either is fine, but we should be consistent.

BatemanVO reviewed (commented):
Very cool applet - and good practice. I believe "to do" lists similar to this one are often presented for simple coding challenges as part of some job interviews, so it's a good one to have under your belt.

aggiewb committed:
Add custom properties for elements that share colors

aggiewb committed:
Replace all variables named ulElement with shoppingList as

suggested in offline review

aggiewb committed:
Wrap file in IIFE and add strict mode

aggiewb committed:
Add map function passing in an anonymous function that trims

each item in the array of its whitespace
Remove for loop conditional as whitespace is removed with .trim() in the array declaration

aggiewb committed:
Replace all single quotes with doubles quotes

aggiewb committed:
Update alt tag for X img that better describes its function

aggiewb committed:
Remove imgElement parameter from appendItemToList() and

call function directly in liElement.appendChild()

aggiewb committed:
Add footer content with spans and remove js setting of

content with innerHTML
Use textContent on "current-year" span to dynamically update year copyright in JS
